File: Count Submatrices With All Ones/Count Submatrices With All Ones.java
// Solution for Count Submatrices With All Ones in JAVA

class Solution {
    private int countSubarraysWithOnes(int[] arr) {
        int ones = 0, total = 0;
        for (int val : arr) {
            if (val == 0) ones = 0;
            else ones++;
            total += ones;
        }
        return total;
    }

    public int numSubmat(int[][] mat) {
        int rows = mat.length, cols = mat[0].length, total = 0;
        int[] mask = new int[cols];
        for (int i = 0; i < rows; i++) {
            Arrays.fill(mask, 1);
            for (int j = i; j < rows; j++) {
                for (int k = 0; k < cols; k++) {
                    mask[k] &= mat[j][k];
                }
                total += countSubarraysWithOnes(mask);
            }
        }
        return total;
    }
}

File: Count Submatrices With All Ones/Count Submatrices With All Ones.py
# Solution for Count Submatrices With All Ones in PY

class Solution:
    def countOnes(self, arr):
        ones, total = 0, 0
        for v in arr:
            if v == 0:
                ones = 0
            else:
                ones += 1
            total += ones
        return total

    def numSubmat(self, mat: List[List[int]]) -> int:
        rows, cols = len(mat), len(mat[0])
        total = 0
        for i in range(rows):
            mask = [1] * cols
            for j in range(i, rows):
                for k in range(cols):
                    mask[k] &= mat[j][k]
                total += self.countOnes(mask)
        return total
